snackpub/src/index.ts

```diff
@@ -0,0 +1,111 @@
+import { Server } from 'socket.io';
+
+import Env from './Env';
+import { maybeBindRedisAdapterAsync, maybeCloseRedisConnectionsAsync } from './RedisAdapter';
+import type {
+  ClientToServerEvents,
+  InterServerEvents,
+  ServerToClientEvents,
+  ShutdownSignal,
+  SocketData,
+} from './types';
+
+const debug = require('debug')('snackpub');
+
+function registerShutdownHandlers(server: Server) {
+  const shutdown = async (signal: ShutdownSignal) => {
+    console.log(
+      `Received ${signal}; the HTTP server is shutting down and draining existing connections`
+    );
+    await closeServerAsync(server);
+    await maybeCloseRedisConnectionsAsync();
+  };
+
+  // TODO: In Node 9, the signal is passed as the first argument to the listener
+  process.once('SIGHUP', () => shutdown('SIGHUP'));
+  process.once('SIGINT', () => shutdown('SIGINT'));
+  process.once('SIGTERM', () => shutdown('SIGTERM'));
+  // Nodemon sends SIGUSR2 when it restarts the process it's monitoring
+  process.once('SIGUSR2', () => shutdown('SIGUSR2'));
+}
+
+function closeServerAsync(server: Server): Promise<void> {
+  return new Promise((resolve) => {
+    server.close((error) => {
+      if (error) {
+        console.log('Failed to close server', error);
+      }
+      resolve();
```

Comment:
Can this hang for some reason? It's always resolving, even when erroring.

Reply:
the function is called in the shutdown handler. after this function finishing its work, we can disconnect redis clients. even if the server closing failed, we still have to disconnect redis clients. that's why it's always resolving here.

other than this approach, we can also do this way, but i think this is more complicated.

```ts
function closeServerAsync(server: Server): Promise<void> {
  return new Promise((resolve, reject) => {
    server.close((error) => {
      if (error) {
        reject(error);
      } else {
        resolve();
      }
    });
  });
}

const shutdown = async (signal: ShutdownSignal) => {
  console.log(
    `Received ${signal}; the HTTP server is shutting down and draining existing connections`
  );
  try {
    await closeServerAsync(server);
  } catch (error) {
    console.log('Failed to close server', error);
  }
  await maybeCloseRedisConnectionsAsync();
};
```